### PR TITLE
Revert "Add blacklister to repository management"

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

In #3047, we made some changes to dependency resolution:

- Add blacklister. If a repository is considered as "corrupted", all subsequent requests sent to it would be ignored.
- Skip resolution in subsequent repositories on failure.

I'd admit this change is way too drastic. "Continue on failure" is sure a bug (see #2853 ), but many builds depend on it. After these changes were shipped in 4.2, we got some bad feedbacks:

- #3335 #3379 #3299 complain about this change broke their builds.
- Twitter experienced an error: twitter's repository didn't response `HEAD` correctly (it should return a `404` but actually threw an exception), so Gradle marked the repo as corrupted, then skipped all subsequent requests. 

In #3374, @bmuschko made an attempt to fix the former issues: do not skip subsequent repositories on failure. However, this is not enough to solve twitter's issue, so this PR reverts #3047 completely. (I had to say sorry to Ben because I mixed the two stories in a single commit so he had to identify the changed files in #3374).

 In addition, to handle timeout issue mentioned in #2853, I add a small change: if timeout occurs, skip the subsequent repositories. Reviewing a subset of this PR may be a bit easier: https://github.com/gradle/gradle/pull/3380/commits/7bba5357d7bd47d3f04f59e44ab433bd459f2398

### CODE REVIEW

**Comment by oehme** [comment]
We don't want to remove blacklisting. We only want to continue on certain errors like 401 or 409. That's what Ben's PR is about. If that doesn't solve some issue, please let Ben know what that issue is.

**Comment by blindpirate** [comment]
The issue is what we talked in slack's `build-tool` channel: twitter's repository would respond Gradle's `HEAD` request with exception (not sure what it is, but not 2xx/3xx/404). This would cause all subsequent requests blacklisted.

**Comment by oehme** [comment]
So the issue is simply that the blacklister stops on any IOException. We have several places in the code base where we throw an IOException instead of a proper Http status code. The blacklister is then accidentally blacklisting those. It should filter by ConnectionTimeout instead.

File: subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpClientHelper.java [hunk 1]
```diff
@@ -183,6 +187,25 @@ public synchronized void close() throws IOException {
         }
     }
 
+    private static boolean isTimeoutException(Throwable throwable) {
+        Throwable rootCause = ExceptionUtils.getRootCause(throwable);
+        //http://hc.apache.org/httpclient-3.x/exception-handling.html
+        return rootCause instanceof ConnectTimeoutException || rootCause instanceof SocketTimeoutException;
```
Comment (bmuschko) [inline subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpClientHelper.java]: I think we should keep the blacklister implementation intact and rather make the check there.

Comment (bmuschko) [reply subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpClientHelper.java]: Ha, never mind. Looks like @bigdaz just did that: https://github.com/gradle/gradle/pull/3381

**Comment by oehme** [comment]
Closing in favor of @bigdaz PRs

**Pull request closed without merging**